Add documentation and tutorial for LlamaIndex Workflow logging (#13336)
mlflow/mlflow: docs/source/llms/llama-index/index.rst

Getting Started
---------------

In these introductory tutorials, you will learn the most fundamental components of LlamaIndex and how to leverage the integration with MLflow to bring better maintainability and observability to your LlamaIndex applications.

.. raw:: html

    <section>
        <article class="simple-grid">
            <div class="simple-card">
                <a href="notebooks/llama_index_workflow_tutorial.html">
                    <div class="header">
                        LlamaIndex Workflows with MLflow
                    </div>
                    <p>
                        Get started with MLflow and LLamaIndex by building a simple agentic Workflow. Learn how to log and load the Workflow for inference, as well as enable tracing for observability.
                    </p>
                </a>
            </div>
            <div class="simple-card">
                <a href="notebooks/llama_index_quickstart.html">
                    <div class="header">

>                         Building Index with MLflow

Reviewer: I think quickstart should be somewhere in the name. I also think this card should be first, as it's the "entry point" to the flavor

Reviewer 2: We're thinking of de-emphasizing the limited scope of the index-focused low-level APIs in favor of the workflow APIs (it seems like this is the direction that the library's going to be taking) so we decided to swap these two with the workflows API taking center stage.

        </article>
    </section>

.. toctree::
    :maxdepth: 2
    :hidden:

    notebooks/llama_index_workflow_tutorial.ipynb
    notebooks/llama_index_quickstart.ipynb

Concepts
--------

``Workflow`` 🆕

> ^^^^^^^^^^^^^^^

Reviewer 2: Could we add a "supported as of llamaindex version 'x'" note (and MLflow version 'x' as well :) ) ?

> The ``Workflow`` is LlamaIndex's new event-driven orchestration framework. It is designed

Reviewer: ```suggestion
The ``Workflow`` is LlamaIndex's event-driven orchestration framework. It is designed
```
`new` will become out of date

> as a highly flexible and interpretable framework for building arbitrary LLM applications such as an agent, a RAG flow, an data extraction pipeline, etc.

Reviewer: ```suggestion
a flexible and interpretable framework for building arbitrary LLM applications such as an agent, a RAG flow, a data extraction pipeline, etc.
```

> MLflow supports tracking, evaluating, and tracing the ``Workflow`` objects, adding strong observability and maintainability values.

Reviewer: ```suggestion
MLflow supports tracking, evaluating, and tracing ``Workflow`` objects, which makes them more observable and maintainable.
```

query and returns a response based on the index. The ``ChatEngine`` is designed for conversational agents, which keeps track of the conversation history as well.

``Settings``
^^^^^^^^^^^^

> The ``Settings`` is a global configuration object in LlamaIndex that bundles commonly used resources throughout the

Reviewer: ```suggestion
The ``Settings`` object is a global service context in LlamaIndex that bundles commonly used resources throughout the
```

> LlamaIndex application. It includes settings such as the LLM model, embedding model, callbacks, and more. MLflow tracks 

Reviewer: ```suggestion
LlamaIndex application. It includes settings such as the LLM model, embedding model, callbacks, and more. When logging a LlamaIndex index/engine/workflow, MLflow tracks 
```

    objects support will be added in the future releases.

How to log and load a LlamaIndex Workflow?
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> Mlflow support logging and loading a LlamaIndex Workflow via the `Model-from-Code <../../../models.html#models-from-code>`_ logging. For the detailed example of logging and loading a LlamaIndex Workflow, see the `LlamaIndex Workflows with MLflow <notebooks/llama_index_workflow_tutorial.html>`_ notebook.

Reviewer 2: ```suggestion
Mlflow supports logging and loading a LlamaIndex Workflow via the `Model-from-Code <../../../models.html#models-from-code>`_ feature. For a detailed example of logging and loading a LlamaIndex Workflow, see the `LlamaIndex Workflows with MLflow <notebooks/llama_index_workflow_tutorial.html>`_ notebook.
```

.. code-block:: python

    import mlflow

    with mlflow.start_run():
        model_info = mlflow.llama_index.log_model(
            "/path/to/workflow.py",
            artifact_path="model",
            input_example={"input": "What is MLflow?"},
        )

The logged workflow can be loaded back using the :py:func:`mlflow.llama_index.load_model` or :py:func:`mlflow.pyfunc.load_model` function.

.. code-block:: python

    # Use mlflow.llama_index.load_model to load the workflow object as is
    workflow = mlflow.llama_index.load_model(model_info.model_uri)
    await workflow.run(input="What is MLflow?")

    # Use mlflow.pyfunc.load_model to load the workflow as a MLflow Pyfunc Model
    # with standard inference APIs for deployment and evaluation.
    pyfunc = mlflow.pyfunc.load_model(model_info.model_uri)
    pyfunc.predict({"input": "What is MLflow?"})

.. warning::

>     The MLflow Pyfunc Model does not support async inference. When you load the workflow with :py:func:`mlflow.pyfunc.load_model`, the ``predict`` method becomes **synchronous** and will block until the workflow execution is completed. This also applies when deploying the logged LlamaIndex workflow to a production endpoint using `MLflow Deployment <https://mlflow.org/docs/latest/deployment/index.html>`_ or Databricks `Model Serving <https://docs.databricks.com/en/machine-learning/model-serving/index.html>`_.

Reviewer: ```suggestion
    The MLflow PyFunc Model does not support async inference. When you load the workflow with :py:func:`mlflow.pyfunc.load_model`, the ``predict`` method becomes **synchronous** and will block until the workflow execution is completed. This also applies when deploying the logged LlamaIndex workflow to a production endpoint using `MLflow Deployment <https://mlflow.org/docs/latest/deployment/index.html>`_ or Databricks `Model Serving <https://docs.databricks.com/en/machine-learning/model-serving/index.html>`_.
```
